### PR TITLE
[Merged by Bors] - Feat: Add some operations and lemmas on closure operators/order homs

### PR DESCRIPTION
**Pull request by Shamrock-Frost**
This adds conjugation of order homomorphisms & closure operators by order isos, as well as two new extensionality lemmas for closure operators, a proof that the inf of a closed family is closed, and that the closure of an element is the GLB of all closed elements larger than it. There is also includes some minor refactoring, moving `Set.image_sSup` from `Mathlib/Order/Hom/CompleteLattice` to `Mathlib/Data/Set/Lattice` and adding some common lemmas for `EquivLike`-things to `OrderIso`.

---

- [x] depends on: #10347

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by Shamrock-Frost** [comment]
I don't remember why there's a dependency on the `Nonempty.imp` PR, it might be that that lemma wasn't used until later in the chain of PRs I'm doing right now. But I already rewrote my git history extensively and I don't want to bother with doing it again when the `Nonempty.imp` PR is so small

File: Mathlib/Order/Hom/Basic.lean [hunk 1]
```diff
@@ -935,6 +935,41 @@ theorem symm_trans (e₁ : α ≃o β) (e₂ : β ≃o γ) : (e₁.trans e₂).s
...
+theorem symm_trans_self (e : α ≃o β) : e.symm.trans e = OrderIso.refl β :=
+  RelIso.symm_trans_self e
+
+-- todo: behavior of this under composition of f, g and of the input
+/-- An order isomorphism between the domains and codomains of two prosets of
+order homomorphisms gives an order isomorphism between the two function prosets. -/
+@[simps apply symm_apply]
+def arrowCongr {α β γ δ} [Preorder α] [Preorder β] [Preorder γ] [Preorder δ]
+    (f : α ≃o γ) (g : β ≃o δ) : (α →o β) ≃o (γ →o δ) where
+  toFun := (.comp g $ .comp . f.symm)
```
Comment (jcommelin) [inline Mathlib/Order/Hom/Basic.lean:952]: Please consider using `<|` (and `|>`).
```suggestion
  toFun := (.comp g <| .comp . f.symm)
```

Comment (Shamrock-Frost) [reply Mathlib/Order/Hom/Basic.lean]: Done

File: Mathlib/Order/Closure.lean [hunk 2]
```diff
@@ -262,6 +298,11 @@ def ofCompletePred (p : α → Prop) (hsinf : ∀ s, (∀ a ∈ s, p a) → p (s
     (fun a ↦ hsinf _ <| forall_range_iff.2 fun b ↦ b.2.2)
     (fun a b hab hb ↦ iInf_le_of_le ⟨b, hab, hb⟩ le_rfl)
 
+theorem sInf_IsClosed {c : ClosureOperator α} {S : Set α}
+    (H : ∀ x ∈ S, c.IsClosed x) : c.IsClosed (sInf S) :=
+  isClosed_iff_closure_le.mpr <| le_of_le_of_eq c.monotone.map_sInf_le <|
+    Eq.trans (biInf_congr (c.isClosed_iff.mp <| H . .)) sInf_eq_iInf.symm
```
Comment (alreadydone) [inline Mathlib/Order/Closure.lean:304]: I think the standard dot for function arguments is the middle dot:
```suggestion
  isClosed_iff_closure_le.mpr <| le_of_le_of_eq c.monotone.map_sInf_le <|
    Eq.trans (biInf_congr (c.isClosed_iff.mp <| H · ·)) sInf_eq_iInf.symm
```
and you could use `le.le_of_eq` and `.trans` dot notations if you like.

File: Mathlib/Order/Hom/Basic.lean [hunk 3]
```diff
@@ -935,6 +935,41 @@ theorem symm_trans (e₁ : α ≃o β) (e₂ : β ≃o γ) : (e₁.trans e₂).s
...
+  RelIso.symm_trans_self e
+
+-- todo: behavior of this under composition of f, g and of the input
+/-- An order isomorphism between the domains and codomains of two prosets of
+order homomorphisms gives an order isomorphism between the two function prosets. -/
+@[simps apply symm_apply]
+def arrowCongr {α β γ δ} [Preorder α] [Preorder β] [Preorder γ] [Preorder δ]
+    (f : α ≃o γ) (g : β ≃o δ) : (α →o β) ≃o (γ →o δ) where
+  toFun := (.comp g <| .comp . f.symm)
+  invFun := (.comp g.symm <| .comp . f)
```
Comment (alreadydone) [inline Mathlib/Order/Hom/Basic.lean:953]:
```suggestion
  toFun := (.comp g <| .comp · f.symm)
  invFun := (.comp g.symm <| .comp · f)
```

File: Mathlib/Order/Closure.lean [hunk 4]
```diff
@@ -195,6 +222,15 @@ theorem IsClosed.closure_le_iff (hy : c.IsClosed y) : c x ≤ y ↔ x ≤ y := b
...
 lemma closure_min (hxy : x ≤ y) (hy : c.IsClosed y) : c x ≤ y := hy.closure_le_iff.2 hxy
 
+lemma closure_IsGLB (x : α) : IsGLB { y | x ≤ y ∧ c.IsClosed y } (c x) where
+  left _ := and_imp.mpr closure_min
+  right _ h := h ⟨c.le_closure x, c.isClosed_closure x⟩
+
+theorem ext_IsClosed (c₁ c₂ : ClosureOperator α)
+    (h : ∀ x, c₁.IsClosed x ↔ c₂.IsClosed x) : c₁ = c₂ :=
+  ext' c₁ c₂ <| fun x => IsGLB.unique (c₁.closure_IsGLB x) <|
+    (Set.ext (and_congr_right' <| h .)).substr (c₂.closure_IsGLB x)
```
Comment (alreadydone) [inline Mathlib/Order/Closure.lean:232]:
```suggestion
    (Set.ext (and_congr_right' <| h ·)).substr (c₂.closure_IsGLB x)
```

File: Mathlib/Order/Closure.lean (hunk 2 above)
Comment (Shamrock-Frost) [reply Mathlib/Order/Closure.lean:304]: Gah, I keep forgetting about the middle dot. When not contributing to mathlib I just use a period for everything. Thanks for the heads up! Also for some reason I really dislike `.trans` over `Eq.trans`

File: Mathlib/Order/Closure.lean [hunk 5]
```diff
@@ -103,9 +125,14 @@ variable {α} [PartialOrder α] (c : ClosureOperator α)
 
 @[ext]
 theorem ext : ∀ c₁ c₂ : ClosureOperator α, (c₁ : α → α) = (c₂ : α → α) → c₁ = c₂ :=
-  DFunLike.coe_injective
+  @DFunLike.ext' _ _ _ _
 #align closure_operator.ext ClosureOperator.ext
 
+-- this should probably be the defautlt?
```
Comment (ocfnash) [inline Mathlib/Order/Closure.lean:131]:
```suggestion
```
It looks to me like you're right. If so, now is the time to enact this! Can you try it (and also have only one lemma carrying `@[ext]`).

File: Mathlib/Order/Closure.lean [hunk 6]
```diff
@@ -195,6 +222,15 @@ theorem IsClosed.closure_le_iff (hy : c.IsClosed y) : c x ≤ y ↔ x ≤ y := b
 
 lemma closure_min (hxy : x ≤ y) (hy : c.IsClosed y) : c x ≤ y := hy.closure_le_iff.2 hxy
 
+lemma closure_IsGLB (x : α) : IsGLB { y | x ≤ y ∧ c.IsClosed y } (c x) where
```
Comment (ocfnash) [inline Mathlib/Order/Closure.lean:225]:
```suggestion
lemma closure_isGLB (x : α) : IsGLB { y | x ≤ y ∧ c.IsClosed y } (c x) where
```

File: Mathlib/Order/Closure.lean [hunk 7]
```diff
@@ -195,6 +222,15 @@ theorem IsClosed.closure_le_iff (hy : c.IsClosed y) : c x ≤ y ↔ x ≤ y := b
...
+  left _ := and_imp.mpr closure_min
+  right _ h := h ⟨c.le_closure x, c.isClosed_closure x⟩
+
+theorem ext_IsClosed (c₁ c₂ : ClosureOperator α)
```
Comment (ocfnash) [inline Mathlib/Order/Closure.lean:229]:
```suggestion
theorem ext_isClosed (c₁ c₂ : ClosureOperator α)
```

File: Mathlib/Order/Closure.lean [hunk 8]
```diff
@@ -262,6 +298,11 @@ def ofCompletePred (p : α → Prop) (hsinf : ∀ s, (∀ a ∈ s, p a) → p (s
     (fun a ↦ hsinf _ <| forall_range_iff.2 fun b ↦ b.2.2)
     (fun a b hab hb ↦ iInf_le_of_le ⟨b, hab, hb⟩ le_rfl)
 
+theorem sInf_IsClosed {c : ClosureOperator α} {S : Set α}
```
Comment (ocfnash) [inline Mathlib/Order/Closure.lean:301]:
```suggestion
theorem sInf_isClosed {c : ClosureOperator α} {S : Set α}
```

File: Mathlib/Order/Hom/Basic.lean [hunk 9]
```diff
@@ -935,6 +935,41 @@ theorem symm_trans (e₁ : α ≃o β) (e₂ : β ≃o γ) : (e₁.trans e₂).s
...
 
+@[simp]
+theorem self_trans_symm (e : α ≃o β) : e.trans e.symm = OrderIso.refl α :=
+  RelIso.self_trans_symm e
+
+@[simp]
+theorem symm_trans_self (e : α ≃o β) : e.symm.trans e = OrderIso.refl β :=
+  RelIso.symm_trans_self e
+
+-- todo: behavior of this under composition of f, g and of the input
```
Comment (ocfnash) [inline Mathlib/Order/Hom/Basic.lean:946]: I think we should either drop this `todo` or make it slightly more precise.

**Review by ocfnash — APPROVED** [review]
Thanks! Please apply and consider the minor suggestions and then feel free to merge.

bors d+

**Pull request closed without merging**